src/download.rs

```diff
@@ -291,6 +285,63 @@ pub fn unpack_bins(dir: &Path, dst_dir: &Path) -> Result<Vec<PathBuf>> {
     Ok(downloaded)
 }
 
+/// write Ok(Response) to provided writer with progress bar displaying writing status
+pub fn write_response_with_progress_bar<W: Write>(
+    response: Response,
+    writer: &mut W,
+) -> Result<()> {
+    let total_size = response
+        .header("Content-Length")
+        .and_then(|s| s.parse::<u64>().ok())
+        .unwrap_or(0);
+    let mut downloaded_size = 0;
+    let mut buffer = [0; 8192];
+    let progress_bar = ProgressBar::new(total_size);
+    progress_bar.set_style(
+                ProgressStyle::default_bar()
+                    .template("[{elapsed_precise}] [{bar:40.cyan/blue}] {bytes}/{total_bytes} {bytes_per_sec} ({eta}) - {msg:.cyan}")
+                    .unwrap()
+                    .progress_chars("##-"),
+            );
+    let mut reader = progress_bar.wrap_read(response.into_reader());
+
+    loop {
+        let bytes_read = reader.read(&mut buffer)?;
+        if bytes_read == 0 {
+            break;
+        }
+        if let Err(e) = writer.write_all(&buffer[..bytes_read]) {
+            debug!(
+                "[{}] [{}] {}/{} {}/s ({}) - {}",
+                FormattedDuration(progress_bar.elapsed()),
+                "#".repeat(
+                    (progress_bar.position() * 40 / progress_bar.length().unwrap()) as usize
+                ),
+                HumanBytes(progress_bar.position()),
+                HumanBytes(progress_bar.length().unwrap_or(progress_bar.position())),
+                HumanBytes(progress_bar.per_sec() as u64),
+                HumanDuration(progress_bar.eta()),
+                progress_bar.message(),
+            );
+            error!("Something went wrong writing data: {}", e)
+        };
+        downloaded_size += bytes_read as u64;
+        progress_bar.set_position(downloaded_size);
+    }
+    progress_bar.finish_with_message("Download complete");
```

Comment:
I think this shows that the download completed even if there was an error. Could you add a unit test where it fails?

Reply:
I have tested in local, when download encount error, won't show "download completed"